gnovm/pkg/gnolang/preprocess.go

```diff
@@ -737,6 +737,75 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 				cx := evalConst(store, last, n)
 				return cx, TRANS_CONTINUE
 
+			// TRANS_LEAVE -----------------------
+			case *FuncLitExpr:
+				// We need to closure-call if any variables
+				// used were declared in a for-loop, and this
+				// is the outer-most function within the loop.
+				// NOTE: if the variable is defined as a
+				// function parameter of the closure, then it
+				// is shadowed and not relevant whether or not
+				// the closure is immediately called.
+
+				// Step 1: check if this func lit is in a loop,
+				// not embedded within another func lit.
+				if !isFuncLitInLoop(store, n) {
+					return n, TRANS_CONTINUE
+				}
+
+				// Step 2: gather all extern names declared in for-loops.
+				leNames := []Name{}
+				leTypes := []Type{}
+				lePaths := []ValuePath{}
+				for name := range n.GetExternNames() {
+					path := n.GetExternPathForName(store, name)
+					if path.Type != VPBlock {
+						continue // not a for-loop block path
+					}
+					// XXX this is wrong, need to check all ancestors
+					// rather than the block node for path.
+					container := getBlockNodeForPath(n, path)
```

Comment:
TODO: refactor isFuncLitInLoop to isNodeInLoop,
and call that for the container too unless it is already a loop.